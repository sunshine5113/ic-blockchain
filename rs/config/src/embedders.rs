use ic_types::NumInstructions;
use serde::{Deserialize, Serialize};

use crate::flag_status::FlagStatus;
use ic_base_types::NumBytes;

// Defining 100000 globals in a module can result in significant overhead in
// each message's execution time (about 40x), so set a limit 3 orders of
// magnitude lower which should still allow for reasonable canisters to be
// written (current max number of globals on the Alpha network is 7).
pub(crate) const MAX_GLOBALS: usize = 300;
// Current max number of functions used by a canister on the Alpha network is
// about 2800, so we set a limit at two times that.
pub(crate) const MAX_FUNCTIONS: usize = 7000;
// The maximum number of custom sections allowed in a wasm module.
pub(crate) const MAX_CUSTOM_SECTIONS: usize = 16;
// The total size of the exported custom sections in bytes.
// The size should not exceed 1MiB.
pub(crate) const MAX_CUSTOM_SECTIONS_SIZE: NumBytes = NumBytes::new(1048576);
/// The number of threads to use for query execution.
pub(crate) const QUERY_EXECUTION_THREADS: usize = 2;

/// In terms of execution time, compiling 1 WASM instructions takes as much time
/// as actually executing 3_000 instructions. Only public for use in tests.
#[doc(hidden)]
const DEFAULT_COST_TO_COMPILE_WASM_INSTRUCTION: NumInstructions = NumInstructions::new(3_000);

#[derive(Clone, Debug, Deserialize, PartialEq, Eq, Serialize)]
pub struct FeatureFlags {
    pub api_cycles_u128_flag: FlagStatus,
    pub rate_limiting_of_debug_prints: FlagStatus,
}

impl Default for FeatureFlags {
    fn default() -> Self {
        Self {
            api_cycles_u128_flag: FlagStatus::Enabled,
            rate_limiting_of_debug_prints: FlagStatus::Enabled,
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
pub struct Config {
    pub max_wasm_stack_size: usize,
    pub query_execution_threads: usize,

    /// Maximum number of globals allowed in a Wasm module.
    pub max_globals: usize,

    /// Maximum number of functions allowed in a Wasm module.
    pub max_functions: usize,

    /// Maximum number of custom sections allowed in a Wasm module.
    pub max_custom_sections: usize,

    /// Maximum size of the custom sections in bytes.
    pub max_custom_sections_size: NumBytes,

    /// Compiling a single WASM instruction should cost as much as executing
    /// this many instructions.
    pub cost_to_compile_wasm_instruction: NumInstructions,

    /// Flags to enable or disable features that are still experimental.
    pub feature_flags: FeatureFlags,
}

impl Config {
    pub fn new() -> Self {
        Config {
            max_wasm_stack_size: 5 * 1024 * 1024,
            query_execution_threads: QUERY_EXECUTION_THREADS,
            max_globals: MAX_GLOBALS,
            max_functions: MAX_FUNCTIONS,
            max_custom_sections: MAX_CUSTOM_SECTIONS,
            max_custom_sections_size: MAX_CUSTOM_SECTIONS_SIZE,
            cost_to_compile_wasm_instruction: DEFAULT_COST_TO_COMPILE_WASM_INSTRUCTION,
            feature_flags: FeatureFlags::default(),
        }
    }
}

impl Default for Config {
    fn default() -> Self {
        Self::new()
    }
}
