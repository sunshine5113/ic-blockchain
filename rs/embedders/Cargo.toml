[package]
name = "ic-embedders"
version = "0.8.0"
edition = "2018"

[dependencies]
anyhow = "1.0.31"
ic-config = { path = "../config" }
ic-cycles-account-manager = { path = "../cycles_account_manager" }
ic-interfaces = { path = "../interfaces" }
ic-logger = { path = "../monitoring/logger" }
ic-registry-subnet-type = { path = "../registry/subnet_type" }
ic-replicated-state = { path = "../replicated_state" }
ic-sys = { path = "../sys" }
ic-system-api = { path = "../system_api" }
ic-metrics = { path = "../monitoring/metrics" }
ic-types = { path = "../types/types" }
ic-utils = { path = "../utils" }
ic-wasm-types = { path = "../types/wasm_types" }
libc = "0.2.91"
libflate = "1.1.2"
memory_tracker = { path = "../memory_tracker" }
nix = "0.23.0"
parity-wasm = { version = "0.42.2", features = [ "std", "multi_value", "bulk" ] }
prometheus = { version = "0.12.0", features = [ "process" ] }
serde = { version = "1.0.99", features = [ "derive" ] }
slog = { version = "2.5.2", features = ["nested-values", "max_level_trace", "release_max_level_debug"] }
wasmtime = { version = "0.37", default_features = false, features = ['cranelift', 'parallel-compilation', 'posix-signals-on-macos', 'wasm-backtrace'] }
wasmtime-environ = "0.37"
wasmtime-runtime = "0.37"


[dev-dependencies]
ic-registry-routing-table = { path = "../registry/routing_table" }
ic-system-api = { path = "../system_api" }
ic-test-utilities = { path = "../test_utilities" }
lazy_static = "1.4.0"
maplit = "1.0.2"
proptest = "0.9.4"
slog = { version = "2.5.2", features = ["nested-values", "max_level_trace", "release_max_level_debug"] }
assert_matches = "1.3.0"
insta = "1.8.0"
pretty_assertions = "0.6.1"
wabt = { git = "https://github.com/dfinity-lab/wabt-rs", tag = "0.10.0-dfinity" }


[[bin]]
name = "instrument-wasm"
path = "src/wasm_utils/instrument_wasm.rs"
