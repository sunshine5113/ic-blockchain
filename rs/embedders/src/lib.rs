mod signal_handler;
pub mod wasm_executor;
pub mod wasm_utils;
pub mod wasmtime_embedder;

use ic_interfaces::execution_environment::ExecutionParameters;
use ic_replicated_state::{ExecutionState, Global, NumWasmPages, PageIndex};
use ic_sys::PageBytes;
use ic_system_api::{sandbox_safe_system_state::SandboxSafeSystemState, ApiType};
use ic_types::{methods::FuncRef, NumBytes};
pub use wasmtime_embedder::{WasmtimeEmbedder, WasmtimeMemoryCreator};

pub struct WasmExecutionInput {
    pub api_type: ApiType,
    pub sandbox_safe_system_state: SandboxSafeSystemState,
    pub canister_current_memory_usage: NumBytes,
    pub execution_parameters: ExecutionParameters,
    pub func_ref: FuncRef,
    pub execution_state: ExecutionState,
}

#[derive(Debug)]
pub struct InstanceRunResult {
    pub dirty_pages: Vec<PageIndex>,
    pub stable_memory_size: NumWasmPages,
    pub stable_memory_dirty_pages: Vec<(PageIndex, PageBytes)>,
    pub exported_globals: Vec<Global>,
}

pub trait LinearMemory {
    fn as_ptr(&self) -> *mut libc::c_void;
}

pub trait ICMemoryCreator {
    type Mem: LinearMemory;

    fn new_memory(
        &self,
        mem_size: usize,
        guard_size: usize,
        instance_heap_offset: usize,
        min_pages: usize,
        max_pages: Option<usize>,
    ) -> Self::Mem;
}
