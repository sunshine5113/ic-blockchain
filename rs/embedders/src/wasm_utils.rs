pub mod decoding;
pub mod errors;
pub mod instrumentation;
pub mod validation;
mod wasm_module_builder;
