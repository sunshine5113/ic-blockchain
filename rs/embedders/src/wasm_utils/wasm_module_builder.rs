use super::validation::RESERVED_SYMBOLS;
use ic_wasm_types::WasmInstrumentationError;
use parity_wasm::builder::{CodeLocation, FunctionDefinition, ModuleBuilder};
use parity_wasm::elements::{ExportEntry, GlobalEntry, Internal, Module};

pub(crate) struct WasmModuleBuilder {
    module_builder: ModuleBuilder,
}

/// A wasm module builder wrapper which validates the input before
/// forwarding it to the parity_wasm::ModuleBuilder.
impl WasmModuleBuilder {
    pub fn new(module_builder: ModuleBuilder) -> Self {
        Self { module_builder }
    }

    pub fn push_function(&mut self, function: FunctionDefinition) -> CodeLocation {
        self.module_builder.push_function(function)
    }

    pub fn push_export(
        &mut self,
        field: &str,
        internal: Internal,
    ) -> Result<u32, WasmInstrumentationError> {
        // Ensures that instrumentation does not accidentally
        // export non-reserved symbols.
        if !RESERVED_SYMBOLS.contains(&field) {
            return Err(WasmInstrumentationError::InvalidExport(format!(
                "Exporting non-reserved symbol {} is not allowed.",
                field
            )));
        }
        Ok(self
            .module_builder
            .push_export(ExportEntry::new(field.to_string(), internal)))
    }

    pub fn with_global(mut self, global: GlobalEntry) -> Self {
        self.module_builder = self.module_builder.with_global(global);
        self
    }

    pub fn build(self) -> Module {
        self.module_builder.build()
    }
}
