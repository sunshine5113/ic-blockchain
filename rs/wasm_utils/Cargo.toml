[package]
name = "ic-wasm-utils"
version = "0.8.0"
edition = "2018"

[dependencies]
ic-wasm-types = { path = "../types/wasm_types" }
parity-wasm = { version = "0.40.0", default-features = true }
wasmtime = { git = "https://github.com/dfinity-lab/wasmtime", rev = "60d36d996dfbd335ac18ee6c36e1f10acfaf0c30", features = ["posix-signals-on-macos"] }

[dev-dependencies]
assert_matches = "1.3.0"
pretty_assertions = "0.6.1"
wabt = "0.10.0"

[[bin]]
name = "instrument-wasm"
path = "src/instrument_wasm.rs"
