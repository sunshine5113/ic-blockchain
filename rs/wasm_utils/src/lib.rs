use wasmtime::Config;

mod errors;
pub mod instrumentation;
pub mod validation;

/// Sets Wasmtime flags to ensure deterministic execution.
pub fn ensure_determinism(config: &mut Config) {
    config
        .wasm_threads(false)
        .wasm_simd(false)
        .cranelift_nan_canonicalization(true);
}
