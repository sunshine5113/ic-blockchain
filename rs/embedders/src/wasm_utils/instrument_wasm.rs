pub mod decoding;
pub mod errors;
pub mod instrumentation;
pub mod validation;
pub mod wasm_module_builder;

use decoding::decode_wasm;
use instrumentation::{instrument, InstructionCostTable, InstrumentationOutput};
use std::sync::Arc;
use validation::validate_wasm_binary;

use ic_config::embedders::Config as EmbeddersConfig;

fn usage() {
    println!(
        r#"
Usage: {} wasm_file

  Validate and instrument the wasm_file and output the result to the stdout."#,
        std::env::current_exe().unwrap().display()
    );
}

fn instrument_wasm(filename: &str) -> std::io::Result<()> {
    use std::io::Write;

    let contents = std::fs::read(filename)?;
    let config = EmbeddersConfig::default();
    let decoded = decode_wasm(Arc::new(contents)).expect("failed to decode canister module");
    if let Err(err) = validate_wasm_binary(&decoded, &config) {
        eprintln!("Failed to validate wasm file {}: {}", filename, err);
        std::process::exit(1);
    }
    match instrument(
        &decoded,
        &InstructionCostTable::default(),
        config.cost_to_compile_wasm_instruction,
    ) {
        Ok(InstrumentationOutput { binary, .. }) => std::io::stdout().write_all(binary.as_slice()),
        Err(err) => {
            eprintln!("Failed to instrument wasm file {}: {}", filename, err);
            std::process::exit(1);
        }
    }
}

fn main() {
    let args: Vec<_> = std::env::args().skip(1).collect();
    let args_refs: Vec<_> = args.iter().map(|s| &s[..]).collect();
    match &args_refs[..] {
        ["--help"] | ["-h"] => {
            usage();
            std::process::exit(0);
        }
        [filename] => {
            if let Err(err) = instrument_wasm(filename) {
                eprintln!("Failed to read {}: {}", filename, err);
                std::process::exit(1);
            }
        }
        _ => {
            usage();
            eprint!("Expected a single argument, got: {:?}", args);
            std::process::exit(1);
        }
    }
}
