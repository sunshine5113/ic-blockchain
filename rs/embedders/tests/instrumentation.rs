use ic_config::embedders::Config as EmbeddersConfig;
use ic_embedders::wasm_utils::{
    instrumentation::{
        export_additional_symbols, instrument, ExportModuleData, InstructionCostTable, Segments,
    },
    validation::RESERVED_SYMBOLS,
};
use ic_sys::{PageIndex, PAGE_SIZE};
use ic_wasm_types::BinaryEncodedWasm;
use insta::assert_snapshot;
use parity_wasm::elements::{self, Module};
use pretty_assertions::assert_eq;
use std::fs;
use wabt::{wat2wasm, Features};

/// Assert what the output of wasm instrumentation should be using the [`insta`]
/// crate.
///
/// When making changes that alter the expected output, changes can be easily reviewed and acked using the [`insta` cli](https://insta.rs/docs/cli/).
/// Expected output is stored in `.snap` files in the `snapshots` folder.
/// When tests fail, the new output will be stored in a `.snap.new` file.
/// Instead of using the `insta` cli, you can review and make changes by
/// directly diffing the `.snap` and `.snap.new` files and save changes by
/// updating the `.snap` file.
fn inject_and_cmp(testname: &str, conf: &InstructionCostTable) {
    let filename = format!(
        "{}/tests/instrumentation-test-data/{}.wat",
        std::env::var("CARGO_MANIFEST_DIR").unwrap(),
        testname
    );
    let content = fs::read_to_string(filename).expect("couldn't read the input file");
    let mut features = Features::new();
    features.enable_bulk_memory();
    let buff = wabt::wat2wasm_with_features(content, features.clone())
        .expect("couldn't convert the input wat to Wasm");
    let output = instrument(
        &BinaryEncodedWasm::new(buff),
        conf,
        ic_config::embedders::Config::default().cost_to_compile_wasm_instruction,
    )
    .expect("couldn't instrument Wasm code");
    let module: Module = parity_wasm::elements::deserialize_buffer(output.binary.as_slice())
        .expect("couldn't deserialize module");
    let out = wabt::wasm2wat_with_features(
        elements::serialize(module).expect("couldn't serialize after metering"),
        features,
    )
    .expect("couldn't convert metered Wasm to wat");
    assert_snapshot!(out);
}

#[test]
fn metering_basic() {
    inject_and_cmp("basic", &InstructionCostTable::new());
}

#[test]
fn metering_basic_import() {
    inject_and_cmp("basic_import", &InstructionCostTable::new());
}

#[test]
fn metering_basic_import2() {
    inject_and_cmp("basic_import2", &InstructionCostTable::new());
}

#[test]
fn metering_basic_import_call() {
    inject_and_cmp("basic_import_call", &InstructionCostTable::new());
}

#[test]
fn metering_element() {
    inject_and_cmp("element", &InstructionCostTable::new());
}

#[test]
fn metering_fac() {
    inject_and_cmp("fac", &InstructionCostTable::new().with_default_cost(2));
}

#[test]
fn metering_recursive() {
    inject_and_cmp("recursive", &InstructionCostTable::new());
}

#[test]
fn metering_app() {
    let conf = InstructionCostTable::new()
        .with_instruction_cost("i32.const".to_string(), 100)
        .with_instruction_cost("local.get".to_string(), 30)
        .with_instruction_cost("i32.xor".to_string(), 2);
    inject_and_cmp("app", &conf);
}

#[test]
fn metering_app2() {
    inject_and_cmp("app2", &InstructionCostTable::new());
}

#[test]
fn metering_start() {
    inject_and_cmp("start", &InstructionCostTable::new());
}

#[test]
fn metering_mixed_imports() {
    inject_and_cmp("mixed_imports", &InstructionCostTable::new());
}

#[test]
fn metering_zero_cost_ops() {
    inject_and_cmp("zero_cost_ops", &InstructionCostTable::new());
}

#[test]
fn metering_control_flow() {
    inject_and_cmp("control_flow", &InstructionCostTable::new());
}

#[test]
fn metering_fizzbuzz() {
    inject_and_cmp("fizzbuzz", &InstructionCostTable::new());
}

#[test]
fn metering_nested_ifs() {
    inject_and_cmp("nested_ifs", &InstructionCostTable::new());
}

#[test]
fn export_mutable_globals() {
    inject_and_cmp("export_mutable_globals", &InstructionCostTable::new());
}

#[test]
fn memory_grow() {
    inject_and_cmp("memory_grow", &InstructionCostTable::new());
}

#[test]
fn simple_loop() {
    inject_and_cmp("simple_loop", &InstructionCostTable::new());
}

#[test]
fn metering_memory_fill() {
    inject_and_cmp("memory_fill", &InstructionCostTable::new());
}

#[test]
fn test_get_data() {
    let output = instrument(
        &BinaryEncodedWasm::new(
            wabt::wat2wasm(
                r#"(module
                (memory 1)
                (data (i32.const 2)  "a tree")
                (data (i32.const 11) "is known")
                (data (i32.const 23) "by its fruit")
            )"#,
            )
            .unwrap(),
        ),
        &InstructionCostTable::new(),
        EmbeddersConfig::default().cost_to_compile_wasm_instruction,
    )
    .unwrap();
    let data = output.data.as_slice();
    assert_eq!((2, b"a tree".to_vec()), data[0]);
    assert_eq!((11, b"is known".to_vec()), data[1]);
    assert_eq!((23, b"by its fruit".to_vec()), data[2]);
    let output = instrument(
        &output.binary,
        &InstructionCostTable::new(),
        EmbeddersConfig::default().cost_to_compile_wasm_instruction,
    )
    .unwrap();
    // the data should have been removed from the instrumented module
    assert_eq!(0, output.data.as_slice().len())
}

#[test]
fn test_chunks_to_pages() {
    let segs = Segments::from(vec![
        (0, vec![1; PAGE_SIZE + 10]), // The segment is larger than a page.
        (PAGE_SIZE + 5, vec![2; 10]), // Overlaps with the segment above.
        (PAGE_SIZE + PAGE_SIZE - 100, vec![3; 200]), // Crosses the page boundary.
    ]);
    let mut pages = segs.as_pages();
    // sort for determinism
    pages.sort_by_key(|p| p.0);
    assert_eq!(pages.len(), 3);
    assert_eq!(pages[0].0, PageIndex::new(0));
    assert_eq!(&pages[0].1[..], &[1; PAGE_SIZE]);
    assert_eq!(pages[1].0, PageIndex::new(1));
    assert_eq!(&pages[1].1[0..5], &[1; 5]);
    assert_eq!(&pages[1].1[5..15], &[2; 10]);
    assert_eq!(&pages[1].1[15..PAGE_SIZE - 100], &[0; PAGE_SIZE - 100 - 15]);
    assert_eq!(&pages[1].1[PAGE_SIZE - 100..PAGE_SIZE], &[3; 100]);
    assert_eq!(pages[2].0, PageIndex::new(2));
    assert_eq!(&pages[2].1[0..100], &[3; 100]);
    assert_eq!(&pages[2].1[100..PAGE_SIZE], &[0; PAGE_SIZE - 100]);
}

#[test]
fn test_exports_only_reserved_symbols() {
    let wasm = wat2wasm(
        r#"
        (module
            (import "ic0" "msg_reply" (func $msg_reply))
                (func $test
                    (call $msg_reply)
                )
        )"#,
    )
    .map(BinaryEncodedWasm::new)
    .unwrap();

    let module = parity_wasm::deserialize_buffer::<Module>(wasm.as_slice()).unwrap();
    let module = export_additional_symbols(module, &ExportModuleData::default()).unwrap();

    let exports = module.export_section().unwrap().entries();
    for export in exports {
        assert!(RESERVED_SYMBOLS.contains(&export.field()))
    }
}
