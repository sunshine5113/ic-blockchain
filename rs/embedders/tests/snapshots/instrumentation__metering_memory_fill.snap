---
source: embedders/tests/instrumentation.rs
expression: out
---
(module
  (type (;0;) (func))
  (type (;1;) (func (param i32 i32) (result i32)))
  (type (;2;) (func (param i32) (result i32)))
  (import "__" "out_of_instructions" (func (;0;) (type 0)))
  (import "__" "update_available_memory" (func (;1;) (type 1)))
  (func (;2;) (type 0)
    global.get 0
    i64.const 4
    i64.sub
    global.set 0
    global.get 0
    i64.const 0
    i64.lt_s
    if  ;; label = @1
      call 0
    end
    i32.const 0
    i32.const 1
    i32.const 16
    call 3
    memory.fill)
  (func (;3;) (type 2) (param i32) (result i32)
    global.get 0
    local.get 0
    i64.extend_i32_u
    i64.lt_s
    if  ;; label = @1
      call 0
    end
    global.get 0
    local.get 0
    i64.extend_i32_u
    i64.sub
    global.set 0
    local.get 0)
  (memory (;0;) 1)
  (global (;0;) (mut i64) (i64.const 0))
  (export "memory" (memory 0))
  (export "canister counter_instructions" (global 0)))

